src/main/java/ch/jalu/configme/migration/VersionMigrationService.java

```diff
@@ -0,0 +1,236 @@
+     *
+     * @param reader the reader with which the configuration file can be read
+     * @param configurationData the configuration data
+     * @return true if a migration has been performed, false otherwise (see constants on {@link MigrationService})
+     */
+    protected boolean performMigrations(@NotNull PropertyReader reader, @NotNull ConfigurationData configurationData) {
+        int currentConfigVersion = readConfigurationVersion(reader, configurationData);
+        AtomicBoolean migrationResult = new AtomicBoolean(NO_MIGRATION_NEEDED);
+
+        // Migrate the configuration from the version 1 to 2 to 3 to ...
+        migrations
+            .forEach(migration -> {
+                if (migration == null) return; // null-safe check.
+
+                int fromVersion = migration.fromVersion();
+                int toVersion = migration.toVersion();
+
+                // start the migration.
+                if (currentConfigVersion == fromVersion) {
+                    migration.migrate(reader, configurationData);
+                    setConfigurationVersion(configurationData, toVersion);
+                    migrationResult.set(MIGRATION_REQUIRED);
+                }
+            });
+
+        return migrationResult.get();
+    }
+
+    /**
+     * @return the nullable config version property comment.
+     */
+    @Nullable
+    protected String getConfigVersionPropertyComment() {
+        throw new IllegalStateException("Not implemented yet");
+        // return null;
+    }
+
+    /**
+     * Gets the current configuration version.
+     * If the property doesn't exist it will be created with the {@link #getStartVersion()}.
+     *
+     * @param reader the reader with which the configuration file can be read
+     * @param configurationData the configuration data
+     * @return the current configuration version.
+     */
+    protected int readConfigurationVersion(@NotNull PropertyReader reader, @NotNull ConfigurationData configurationData) {
+        Property<Integer> configVersionProperty = PropertyInitializer.newProperty(getVersionPropertyName(), getStartVersion());
+
+        // if the version property does not exist just add it.
+        if (!reader.contains(getVersionPropertyName())) {
+            configurationData.setValue(configVersionProperty, configVersionProperty.getDefaultValue());
+            // TODO: add the property comment by using the #getConfigVersionPropertyComment() method.
+        }
+
+        return configVersionProperty.determineValue(reader).getValue();
+    }
+
+    protected void setConfigurationVersion(@NotNull ConfigurationData configurationData, int newVersion) {
+        Property<Integer> configVersionProperty = PropertyInitializer.newProperty(getVersionPropertyName(), newVersion);
+        configurationData.setValue(configVersionProperty, configVersionProperty.getDefaultValue());
+    }
+
+    public interface Migration {
```

Comment:
What do you think about adding this is as a top-level class to this package? Maybe with name `VersionMigration`. 

Since people will need to implement this interface, potentially multiple times, it will be less annoying if this is not a nested class :grin: